fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/AsyncLoadingCache.java

```diff
@@ -54,22 +54,31 @@ public AsyncLoadingCache(long refreshTimeMillis, long deadlineTimeMillis) {
 
     /**
      * If the cache does not contain an entry for <code>key</code>, retrieve the value using the provided asynchronous
-     * {@link Supplier}. The future returned by the supplier is cached immediately, this is to prevent concurrent
-     * attempts to get the value from doing unnecessary (and potentially expensive) work. If the returned asynchronous
-     * operation completes exceptionally, the key will be cleared.
+     * {@link Supplier}. If the value is not currently cached, then the {@code Supplier} is used to load the value
+     * asynchronously. If multiple callers ask for the same key at the same time, then they might duplicate
+     * each other's work in that both callers will result in a future being created. Whichever future completes first
+     * will insert its value into the cache, and all callers that then complete successfully are guaranteed to see that
+     * object until such time as the value is expired from the cache.
      *
      * @param key the key in the cache to lookup
      * @param supplier an asynchronous operation to retrieve the desired value if the cache is empty
      *
      * @return a future containing either the cached value or the result from the supplier
      */
+    @Nonnull
     public CompletableFuture<V> orElseGet(@Nonnull K key, @Nonnull Supplier<CompletableFuture<V>> supplier) {
         try {
-            return cache.get(key, () -> MoreAsyncUtil.getWithDeadline(deadlineTimeMillis, supplier)).whenComplete((ignored, e) -> {
-                if (e != null) {
-                    cache.invalidate(key);
-                }
-            });
+            V cachedValue = cache.getIfPresent(key);
+            if (cachedValue == null) {
+                return MoreAsyncUtil.getWithDeadline(deadlineTimeMillis, supplier).thenApply(value -> {
+                    // Only insert the computed value into the cache if a concurrent caller hasn't.
+                    // Return the value that wound up in the cache.
+                    final V existingValue = cache.asMap().putIfAbsent(key, value);
```

Comment:
I think we might want to handle the situation where value is null a little more carefully. This is not an issue for `LocatableResolver` but could be in the future. If I were trying to cache a nullable value I would be invoking the async `supplier` on every read. Maybe we could cache `value` as an `Optional`? That way we could distinguish:
- The optional is null: no async operation has completed and populated the cache
- The optional is empty: an async operation completed with `null` as the result

Reply:
Good catch. I switched to using a `Cache<K, Optional<V>>` internally and added a test where `null` values are cached.